### PR TITLE
Optimize event target memory layout: allocate in heap

### PR DESCRIPTION
**Pull request by PatriceJiang**
Re: #

### Changelog

* Move event-target listeners to heap

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/core/event/EventTarget.h [hunk 1]
```diff
@@ -545,6 +572,29 @@ class EventTargetImpl final {
     TargetEventIdType _handlerId{1};
     std::array<int, EventCount> _emittingEvent{0};
 };
+
+template <typename T>
+class HeapObject final {
+public:
+    HeapObject() {
+        _data = new T;
```
Comment (minggo) [inline native/cocos/core/event/EventTarget.h]: Please use ccnew

**Review by PatriceJiang — COMMENTED** [review]
fix review

**Review by PatriceJiang — COMMENTED** [review]
review fix

File: native/cocos/core/event/EventTarget.h (hunk 1 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventTarget.h]: fixed

**Review by SantyWang — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**